### ISSUE TITLE
Wrong abi in xml file

### ISSUE THREAD

**Issue by BerndCirotzki** [jongough/testplugin_pi#388]
in the plugin xml file in build directory is :
<target>msvc-wx</target>
correct is :
<target>msvc-wx32</target>

Windows build

**Comment by rgleason** [comment]
Bernd, I don't know why this has happened with the primary testplugin_pi repository (Jon's Frontend2 configuration is only slightly different than the one I have been using for all the TP plugins).  
Thank you for posting this.  As a temporary alternative only, you could look at the Frontend2 configuration in
https://github.com/rgleason/testplugin_pi   updated on Dec 23, 2023.  Hope that is a help.  Hopefully Jon will address this problem.

**Comment by jongough** [comment]
The current version, 1.0.294.0, has the following in the metadata.xml file:
```
<?xml version="1.0" encoding="UTF-8"?>
<plugin version="1">
<name> Testplugin </name>
<version> 1.0.294.0 </version>
<release> 0 </release>
<summary> Plugin to test examples of the ODAPI and JSON interface for ODRAW </summary>

<api-version> 1.17 </api-version>
<open-source> yes </open-source>
<author> Jon Gough </author>
<source> https://github.com/jongough/testplugin_pi </source>

<description>
testplugin Plugin is used to test out the ODraw API and demonstrate how to use it successfully from another plugin
</description>

<target>msvc-wx32</target>
<build-target>msvc</build-target>
<build-gtk></build-gtk>
<target-version>10.0.20348</target-version>
<target-arch>x86</target-arch>
<tarball-url>
https://dl.cloudsmith.io/public/opencpn/testplugin-alpha/raw/names/testplugin_pi-1.0.294.0-msvc-x86-wx32-10.0.20348-MSVC-tarball/versions/1.0.294.0+12995.2507e4a/testplugin_pi-1.0.294.0-msvc-x86-wx32-10.0.20348-MSVC.tar.gz
</tarball-url>
<info-url> https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:opencpn_user_manual:plugins:safety:odraw1.6_pi </info-url>
</plugin>
```
Not sure what is wrong with it. It seems to be correct. The filename is not used in any validation and can be whatever you or I name it.  The important bit is the `<target>msvc-wx32</target>`.

**Comment by BerndCirotzki** [comment]
Sorry, it could be the you missundersood what I meant.
I was not downloading your tarball from cloudsmith.
The following :

I'm compiling your code local on my PC.
Windows10, VS2022, wx3.2.4

compiling the code with VS2022. (release)
then Package->build only package 
then there are the files :
build/testplugin_pi-1.0.294.0-msvc-x86-10.0.19045.xml     (this should be renames to "metadata.xml")
in this file is not "<target>msvc-wx32</target>" there is "<target>msvc-wx</target>"
...
and
/build/testplugin_pi-1.0.294.0-msvc-x86-10.0.19045.tar.gz  (this is the file to import in Opencpn)
not    "testplugin_pi-1.0.294.0-msvc-x86-10.0.19045-MSVC.tar.gz"  and in this file there is missing the "metadata.xml"

so "OpenCPN" is not accepting the import.
.... I found this because I would take your code as framework for building a own Plugin.

**Comment by rgleason** [comment]
Bernd, you should use a bash prompt after building to execute the "bld.bat" batch file in my rwpos. Before doing so, check and fix the pathnames to you wxwidets dir.  This will put the necessary metedata.xml file in your tarball.

You cannot use the MS command prompt for this, but it does build the tarball w/o executing the upload.sh script.

**Comment by jongough** [comment]
bld.bat is available in the testplugin root directory.

**Comment by BerndCirotzki** [comment]
Yes! But not working
not
with "LOCAL_BUILD=false" and also not  with  "LOCAL_BUILD=true" in cloudsmith.sh file

**Comment by BerndCirotzki** [comment]
no metadata.xml in tarball.

**Comment by jongough** [comment]
Because the build process is for the cloud with the installables sent to the cloud. If you want to build it locally it will build, if you want to create a local installable file you have to follow the build process and use what is generated. The cloudsmith_upload.sh (generated during the build process from an '*.in" file) creates the final package.

If you are debugging then you only need the library and copy it to the location that opencpn is using. I don't have a local build environment for windows as it just does not seem to work for me. Rick uses windows all the time and he builds OK with it.

Unless you have your own cloudsmith repository or write access to the ones used for the build cycle you will not 'upload' anything as that will fail. However, the tar file should contain what you would find in the cloudsmith repository.

Basically you have to simulate what the circleci build process does, i.e. create the required environment variables, have a version of MSVC 2022 installed, then run the "circleci-build-msvc.bat" file, then run the cloudsmith-upload.sh with the "CIRCLECI_LOCAL=true" environment variable set. This will create the final tarball with the correct contents.

What are you trying to achieve by creating a local version of the tarball?

**Comment by jongough** [comment]
I have just copied testplugin_pi to a windows 10 virtual machine, opened a cmd.exe prompt, gone into the testplugin_pi directory, created the `build` directory, changed into it and issued `cmake -T v143 -A win32 -DOCPN_TARGET=MSVC ..`, then issued `cmake --build . --target package --config relwithdebinfo > output.txt`.
This all worked. The next step I used windows `git bash`, changed into the build directory and issued `bash ./cloudsmith-upload.sh`. This generated the `testplugin_pi**.tar.gz` file and the `testplugin_pi**.xml` file. 

The xml file contained the following:
```
<?xml version="1.0" encoding="UTF-8"?>
<plugin version="1">
<name> Testplugin </name>
<version> 1.0.294.0 </version>
<release> 0 </release>
<summary> Plugin to test examples of the ODAPI and JSON interface for ODRAW </summary>

<api-version> 1.17 </api-version>
<open-source> yes </open-source>
<author> Jon Gough </author>
<source> https://github.com/jongough/testplugin_pi </source>

<description>
testplugin Plugin is used to test out the ODraw API and demonstrate how to use it successfully from another plugin
</description>

<target>msvc-wx32</target>
<build-target>msvc</build-target>
<build-gtk></build-gtk>
<target-version>10.0.19045</target-version>
<target-arch>x86</target-arch>
<tarball-url>
https://dl.cloudsmith.io/public/opencpn/testplugin-alpha/raw/names/testplugin_pi-1.0.294.0-msvc-x86-wx32-10.0.19045-tarball/versions/1.0.294.0+1.5c0ed70/testplugin_pi-1.0.294.0-msvc-x86-wx32-10.0.19045-MSVC.tar.gz
</tarball-url>
<info-url> https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:opencpn_user_manual:plugins:safety:odraw1.6_pi </info-url>
</plugin>
```
and the tar.gz file contained the metadata.xml which had in it:
```
<?xml version="1.0" encoding="UTF-8"?>
<plugin version="1">
<name> Testplugin </name>
<version> 1.0.294.0 </version>
<release> 0 </release>
<summary> Plugin to test examples of the ODAPI and JSON interface for ODRAW </summary>

<api-version> 1.17 </api-version>
<open-source> yes </open-source>
<author> Jon Gough </author>
<source> https://github.com/jongough/testplugin_pi </source>

<description>
testplugin Plugin is used to test out the ODraw API and demonstrate how to use it successfully from another plugin
</description>

<target>msvc-wx32</target>
<build-target>msvc</build-target>
<build-gtk></build-gtk>
<target-version>10.0.19045</target-version>
<target-arch>x86</target-arch>
<tarball-url>
https://dl.cloudsmith.io/public/opencpn/testplugin-alpha/raw/names/testplugin_pi-1.0.294.0-msvc-x86-wx32-10.0.19045-tarball/versions/1.0.294.0+1.5c0ed70/testplugin_pi-1.0.294.0-msvc-x86-wx32-10.0.19045-MSVC.tar.gz
</tarball-url>
<info-url> https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:opencpn_user_manual:plugins:safety:odraw1.6_pi </info-url>
</plugin>
```
and the folder containing all the plugin contents.

So I am not really sure what is missing.

**Comment by BerndCirotzki** [comment]
... this is my xml : (did the same)
<?xml version="1.0" encoding="UTF-8"?>
<plugin version="1">
  <name> Testplugin </name>
  <version> 1.0.294.0 </version>
  <release> 0 </release>
  <summary> Plugin to test examples of the ODAPI and JSON interface for ODRAW </summary>

  <api-version> 1.17 </api-version>
  <open-source> yes </open-source>
  <author> Jon Gough </author>
  <source> https://github.com/ </source>

  <description>
    testplugin Plugin is used to test out the ODraw API and demonstrate how to use it successfully from another plugin
  </description>

  <target>msvc-wx</target>
  <build-target>msvc</build-target>
  <build-gtk></build-gtk>
  <target-version>10.0.19045</target-version>
  <target-arch>x86</target-arch>
  <tarball-url>
    https://dl.cloudsmith.io/public/--pkg_repo--/raw/names/--name--/versions/--version--/--filename--
  </tarball-url>
  <info-url> https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:opencpn_user_manual:plugins:safety:odraw1.6_pi </info-url>
</plugin>

... No metadata ! in tar.gz file

**Comment by BerndCirotzki** [comment]
and the cloudshith upload : 

Home PC@DESKTOP-65G3VNL MINGW64 /c/Entwicklung/testplugin_pi/build
$ ./cloudsmith-upload.sh
+ PROD_REPO=opencpn/testplugin-prod
+ BETA_REPO=opencpn/testplugin-beta
+ ALPHA_REPO=opencpn/testplugin-alpha
+ LOCAL_BUILD=false
+ '[' '' ']'
+ '[' '' ']'
+ '[' '' ']'
+ BUILD_ID=1
+ BUILD_DIR=.
+ BUILD_BRANCH=
+ BUILD_TAG=
+ PKG_EXT=deb
+ LOCAL_BUILD=true
+ set +x
+ '[' '' ']'
+ '[' '' ']'
++ git rev-parse --short=7 HEAD
fatal: not a git repository (or any of the parent directories): .git
+ commit=
+ commit=unknown
++ git tag --contains HEAD
fatal: not a git repository (or any of the parent directories): .git
+ tag=

**Comment by BerndCirotzki** [comment]
But however I don't want make anything with cloudsmith !!
I only want  local a working  xxx.tar.gz file that I can insert manuel into OPENCPN

**Comment by jongough** [comment]
As you are running it locally it will not go to cloudsmith, but the code to create the XML and imbed it in the tar file will be executed. The upload will not happen. You do have to have a clone of the git repository and you will need got functional on your windows machine.

**Comment by BerndCirotzki** [comment]
> You do have to have a clone of the git repository and you will need got functional on your windows machine.
Now I think you understand the problem ... the is realy that, I don't want to do !!! I am on the way, developing a new Plugin, first only for me, up to the moment, when it's working fine, I could make in public on Github.  But that is far away.
So I take your "framwork" as it is described to use it. Everything is working fine, only I cound not insert it into OpenCPN.
... so I found out the reason it is so ! And that is what I wrote here as this Issure !

Let me say, I know now how to fix it for me manual ... that's not the probem.
But I think this could not be the way to do !!!!!
This appears for me as the is a small group of developer how only accept "developing over cloudsmith and github".

I only want to develope me a plugin local, insert in O and ready.

When it is so, than this is not an issure to "your Plugin" it is  an CONCEPT ERROR of opencpn.
Sorry for writing this.

**Comment by BerndCirotzki** [comment]
I think this can be closed now.

**Comment by jongough** [comment]
This plugin is meant to be used as an example or the start of a new plugin with all the bits working. The concept is used by quite a few plugins and does all the work of building for all the supported environments.

With the use of plugin manager the developer process is more complex, but the user experience is easier.

Good luck with your plugin

**Comment by rgleason** [comment]
Bernd,
I would be interested in adding a few lines to the "bld.bat" script that copies metadata.xml into the the tarball. That would solve the problem for windows devs and we could just use MS command. Any ideas how to do that?

**Comment by rgleason** [comment]
Something like:
```
copy [tactics_pi]*-MSVC.xml to metadata.xml
7z a -tzip  [plugin]*.tar.gz  metadata.xml

```
Or perhaps, recreate the entire tar.gz file?

===============
`7z a -tzip archive.zip subdir\`
adds all files and subfolders from folder subdir to archive archive.zip.

`7z a -t7z Files.7z *.txt -r`
adds all *.txt files from current folder and its subfolders to archive Files.7z.

**Comment by BerndCirotzki** [comment]
yes, thanks . I know how do it.

01.03.2024 15:52:35 Rick Gleason ***@***.***>:

**Comment by jongough** [comment]
cloudsmith-upload.sh updates the xml file with the required settings and is a required component to generate plugin manager installable packages. If run locally it just leaves them in the build directory. If run as part of 'circleci local' it will copy them in the designated output directory. If run on circleci it will copy them to the correct cloudsmith repository.

Unless you are testing plugin manager the normal way is to move the plugin file to the requisite lib directory for OCPN to find and run it that way. Going through plugin manager just slows down the testing process.

**Comment by rgleason** [comment]
Ok, Jon so you are suggesting to simply add to the bld.bat some lines that copy 

from ..[plugin]\build\relwithdebinfo\[plugin].dll to C:\Users\fcgle\source\repos\opencpn\plugins
from  ..[plugin]\build\relwithdebinfo\[plugin].pdb  to   C:\Users\fcgle\source\repos\opencpn\plugins

Which is where I run the Development version of OpenCPN  with MS Visual Studio?

**Comment by jongough** [comment]
When I do a Debug build in VS2022 it puts the plugin dll, lib and exp files in the 'build-directory'/Debug directory. These would need to be copied to the 'plugins' subdirectory of the 'Debug' directory. Different builds will have different locations.

If you want to you could change the output directory for the plugin build to be the correct place. I think this may be done in the solution information. But don't take my word for that as I am not a windows user/builder.
